src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs

```diff
@@ -87,63 +96,91 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
             {
-                return;
-            }
 
-            // If we can't figure out the type being created, or the type isn't in source,
-            // then there's nothing we can do.
-            var type = semanticModel.GetSymbolInfo(typeNode, cancellationToken).GetAnySymbol() as INamedTypeSymbol;
-            if (type == null)
-            {
-                return;
+                // Not supported if this is "new { ... }" (as there are no parameters at all.
+                var typeNode = syntaxFacts.GetObjectCreationType(objectCreation);
+                if (typeNode == null)
+                {
+                    return new RegisterFixData<TArgumentSyntax>();
+                }
+
+                // If we can't figure out the type being created, or the type isn't in source,
+                // then there's nothing we can do.
+                if (!(semanticModel.GetSymbolInfo(typeNode, cancellationToken).GetAnySymbol() is INamedTypeSymbol type))
+                {
+                    return new RegisterFixData<TArgumentSyntax>();
+                }
+
+                if (!type.IsNonImplicitAndFromSource())
+                {
+                    return new RegisterFixData<TArgumentSyntax>();
+                }
+
+                var arguments = (SeparatedSyntaxList<TArgumentSyntax>)syntaxFacts.GetArgumentsOfObjectCreationExpression(objectCreation);
+                var methodCandidates = type.InstanceConstructors;
+
+                return new RegisterFixData<TArgumentSyntax>(arguments, methodCandidates);
             }
 
-            if (!type.IsNonImplicitAndFromSource())
+            return null;
+        }
+
+        /// <summary>
+        /// In VB a constructor calls other constructor overloads via a Me.New(..) invocation. We need to exclude the constructor with
+        /// the diagnostic from the candidates because that would likely introduces a recursive call to itself, which in turn would be
+        /// compiler error BC30298: Constructor 'Public Sub New(...)' cannot call itself.
+        /// </summary>
+        private static ImmutableArray<IMethodSymbol> RemoveConstructorWithDiagnosticFromCandidates(
+            TInvocationExpressionSyntax invocationExpression,
+            ImmutableArray<IMethodSymbol> candidates,
+            CancellationToken cancellationToken)
+        {
+            foreach (var candidate in candidates)
             {
-                return;
-            }
+                if (candidate.MethodKind != MethodKind.Constructor)
+                {
+                    break;
+                }
 
-            var arguments = (SeparatedSyntaxList<TArgumentSyntax>)syntaxFacts.GetArgumentsOfObjectCreationExpression(objectCreation);
-            var methodCandidates = type.InstanceConstructors;
+                if (candidate.DeclaringSyntaxReferences.Any(syntaxReference
+                    => syntaxReference.GetSyntax(cancellationToken)?.Parent?.Contains(invocationExpression) == true))
```

Comment:
i really dislike how this reads :)   It's really ok to break things up into multiple statements :)

Reply:
Note: don't you want to do this removal not just for the invocation-expr case?  i.e. in C# if you do `: this(...)` then we also don't want to suggest adding the parameter to the containing constructor.  Can we unify these codepaths?

Reply:
> Can we unify these codepaths?

I tried, but I couldn't come up with a good solution. I tried:

1. Take a SyntaxNode from the method to remove from the candidates

```CS
RemoveConstructorWithDiagnosticFromCandidates(
            SyntaxNode nodeInMethodToRemoveFromCandidates,
            ImmutableArray<IMethodSymbol> candidates,
            CancellationToken cancellationToken)
```
This does not work, because in VB you need go to the parent of the declaration and call `contains` on it:
`syntaxReference.GetSyntax(cancellationToken)?.Parent?.Contains(nodeInMethodToRemoveFromCandidates) == true)`

while in C# the method body is included (and calling `Parent` give you the whole type):

`syntaxReference.GetSyntax(cancellationToken)?.Contains(nodeInMethodToRemoveFromCandidates) == true)`

2. Take the IMethodSymbol of the candidate to remove
```CS
RemoveConstructorWithDiagnosticFromCandidates(
            IMethodSymbol methodToRemove,
            ImmutableArray<IMethodSymbol> candidates)
{
  return (methodToRemove.MethodKind == MethodKind.Constructor)
    ? candidates.Remove(methodToRemove)
    : candidates;
}
```

Now the whole method has degenerated and also the `RegisterFixData` type needs a new property `MethodToRemove` (making the type somewhat stupid having the properties `MethodCandidates` and `MethodToRemove` :dizzy_face:).

Both attempts complicate things and that's why I would prefer to keep the two code paths separated. But maybe you have another idea on how to solve this problem.